### PR TITLE
Add bounded-memory training and colocation controls

### PR DESCRIPTION
**Pull request by YJHMITWEB**
# Bounded-Memory Megatron Training and Colocation

## Summary

This PR adds controls that bound temporary training memory and make colocated Megatron offload behavior complete.

## Changes

| Change | Explanation |
| --- | --- |
| CPU-resident policy microbatches | Keeps the DP-sharded policy batch on CPU and transfers one nested microbatch to CUDA immediately before its forward step. |
| Bounded vocabulary entropy | Chunks Megatron vocabulary entropy by an explicit size or an automatically calculated memory budget, and skips fully masked chunks. |

CPU-resident policy microbatches is opt-in; automatic vocabulary-entropy chunk sizing, which defaults to a 512 MiB temporary-memory budget.

## Memory Validation

| Variant | Memory Spike | Outcome |
| --- | ---: | --- |
| Earlier configuration | 9.3 GiB | OOM easily as the context/vocab size gets larger |
| Bounded-memory configuration | 0.5 GiB | All memory spikes are under control |

### CODE REVIEW

File: skyrl/backends/skyrl_train/workers/worker.py [hunk 1]
```diff
@@ -243,9 +245,86 @@ def init_model(self, *args, **kwargs):
...
+                gpu_uuid = raw_gpu_uuid.decode("ascii")
+            else:
+                gpu_uuid = str(raw_gpu_uuid) if raw_gpu_uuid else None
+            pynvml.nvmlInit()
+            try:
+                handle = (
+                    pynvml.nvmlDeviceGetHandleByUUID(gpu_uuid)
+                    if gpu_uuid
+                    else pynvml.nvmlDeviceGetHandleByIndex(device)
+                )
```
Comment (jinghanyao1-hub) [inline skyrl/backends/skyrl_train/workers/worker.py]: fixed

File: skyrl/backends/skyrl_train/distributed/megatron/model_utils.py [hunk 2]
```diff
@@ -1342,4 +1389,38 @@ def vocab_parallel_entropy(vocab_parallel_logits: torch.Tensor) -> torch.Tensor:
...
+
+    # Keep an autograd edge even when every chunk is masked out.
+    local_entropy_sum = vocab_parallel_logits[..., :0, :].sum()
+    seq_len = int(vocab_parallel_logits.shape[-2])
+    for start in range(0, seq_len, resolved_chunk_size):
+        end = min(start + resolved_chunk_size, seq_len)
+        weight_chunk = weights[start:end]
+        if torch.count_nonzero(weight_chunk).item() == 0:
+            continue
+        entropy_chunk = _VocabParallelEntropy.apply(vocab_parallel_logits[..., start:end, :]).squeeze(0)
```
Comment (jinghanyao1-hub) [inline skyrl/backends/skyrl_train/distributed/megatron/model_utils.py]: fixed

File: skyrl/backends/skyrl_train/workers/megatron/megatron_model_wrapper.py [hunk 3]
```diff
@@ -325,6 +345,8 @@ def collection_func(logits, data):
 
         def forward_step(batch_iter, model):
             batch = next(batch_iter)
+            if self.cpu_resident_microbatches:
```
Comment (erictang000) [inline skyrl/backends/skyrl_train/workers/megatron/megatron_model_wrapper.py]: can we get rid of this config flag and just make this the default behavior? (i.e. only put microbatches onto device when needed?) this seems reasonable to me, overhead probably low anyway

File: skyrl/backends/skyrl_train/workers/worker.py [hunk 4]
```diff
@@ -567,6 +568,9 @@ def __init__(
         self.colocate_all = colocate_all
         self.sequence_parallel_size = sequence_parallel_size
         self.record_memory = record_memory
+        self._pg = pg
+        self._num_gpus_per_actor = num_gpus_per_actor
```
Comment (erictang000) [inline skyrl/backends/skyrl_train/workers/worker.py]: i think these don't get used anywhere else? can remove the above 2 lines?

File: skyrl/backends/skyrl_train/distributed/megatron/model_utils.py [hunk 5]
```diff
@@ -1342,4 +1393,38 @@ def vocab_parallel_entropy(vocab_parallel_logits: torch.Tensor) -> torch.Tensor:
...
+
+    entropy_chunks = []
+    seq_len = int(vocab_parallel_logits.shape[-2])
+    for start in range(0, seq_len, resolved_chunk_size):
+        end = min(start + resolved_chunk_size, seq_len)
+        entropy_chunks.append(_VocabParallelEntropy.apply(vocab_parallel_logits[..., start:end, :]))
+    return torch.cat(entropy_chunks, dim=-1)
+
+
+def vocab_parallel_entropy_weighted_sum(
```
Comment (erictang000) [inline skyrl/backends/skyrl_train/distributed/megatron/model_utils.py:1408]: just to confirm these changes are stable, can you run a basic gsm8k example before and after this PR, and link it here? Just want to check there is no significant speed regression from chunking entropy calculation, and that the entropy metric is exactly matching before

**Review by erictang000 — COMMENTED** [review]
looks good, just some minor comments

File: skyrl/backends/skyrl_train/distributed/megatron/model_utils.py (hunk 5 above)
Comment (jinghanyao1-hub) [reply skyrl/backends/skyrl_train/distributed/megatron/model_utils.py:1408]: We have a unit test on this part: `tests/backends/skyrl_train/distributed/test_vocab_entropy_chunking.py`, this verifies that the chunked CE is numerically identical. For the overhead, the following table shows:

| `chunk_memory_mb` | tokens/chunk | #chunks | exact | op time | vs unchunked | 
| ---: | ---: | ---: | :---: | ---: | ---: | 
| off (unchunked) | — | — | — | 6.3 ms | 1.0× | 
| 4096 | 4096 | 2 | ✓ | 6.4 ms | 1.0× | 
| 2048 | 2048 | 4 | ✓ | 7.5 ms | 1.2× | 
| 1024 | 1024 | 8 | ✓ | 14.9 ms | 2.4× | 
| 512 *(current default)* | 512 | 16 | ✓ | 29.7 ms | 4.7× | 
| 256 | 256 | 32 | ✓ | 59.6 ms | 9.4× | 
| 128 | 128 | 64 | ✓ | 118.6 ms | 18.8× | 

The slowdown is expected, but since the overall step time here is at most hundreds of ms, the impact of chunking is almost negligible.

File: skyrl/backends/skyrl_train/workers/worker.py (hunk 4 above)
Comment (jinghanyao1-hub) [reply skyrl/backends/skyrl_train/workers/worker.py]: fixed

File: skyrl/backends/skyrl_train/workers/megatron/megatron_model_wrapper.py (hunk 3 above)
Comment (jinghanyao1-hub) [reply skyrl/backends/skyrl_train/workers/megatron/megatron_model_wrapper.py]: fixed

**Review by erictang000 — APPROVED** [review]
nice, looking good thanks!
